src/note_encryption/orchard_domain.rs

```diff
@@ -0,0 +1,61 @@
+use core::fmt;
+
+use zcash_note_encryption_zsa::{AEAD_TAG_SIZE, MEMO_SIZE};
+
+use crate::{action::Action, note::Nullifier, Note};
+
+use super::{
+    note_bytes::{NoteByteReader, NoteByteWriter},
+    Memo,
+};
+
+/// Represents the Orchard protocol domain specifics required for note encryption and decryption.
+pub trait OrchardDomain: fmt::Debug + Clone {
+    /// The size of a compact note, specific to the Orchard protocol.
+    // 52 for Vanuilla, 84 for ZSA
+    const COMPACT_NOTE_SIZE: usize;
+
+    /// The size of a note plaintext, including memo and other metadata.
+    // + 512 (564 for Vanilla, 596 for ZSA)
+    const NOTE_PLAINTEXT_SIZE: usize = Self::COMPACT_NOTE_SIZE + MEMO_SIZE;
+
+    /// The size of an encrypted note ciphertext, accounting for additional AEAD tag space.
+    // + 16 (580 for Vanilla, 612 for ZSA)
+    const ENC_CIPHERTEXT_SIZE: usize = Self::NOTE_PLAINTEXT_SIZE + AEAD_TAG_SIZE;
+
+    /// A type to represent the raw bytes of a note plaintext.
+    type NotePlaintextBytes: NoteByteWriter;
+    /// A type to represent the raw bytes of an encrypted note plaintext.
+    type NoteCiphertextBytes: NoteByteReader;
+    /// A type to represent the raw bytes of a compact note.
+    type CompactNotePlaintextBytes: NoteByteWriter;
+    /// A type to represent the raw bytes of an encrypted compact note.
+    type CompactNoteCiphertextBytes: NoteByteReader;
+
+    /// Builds NotePlaintextBytes from Note and Memo.
+    fn build_note_plaintext_bytes(note: &Note, memo: &Memo) -> Self::NotePlaintextBytes;
+}
+
+/// Orchard-specific note encryption logic.
+#[derive(Debug, Clone)]
+pub struct OrchardDomainBase<D: OrchardDomain> {
```

Comment:
From @PaulLaux:
let's rename to EncDomainBase

Reply:
In our discussion, we decided to keep the name `OrchardDomain` for now (along with `OrchardDomainBase`, `OrchardDomainVanilla`, and `OrchardDomainZSA`) because we have three traits - `OrchardDomain`, `OrchardCircuit`, and `OrchardHash`. This naming scheme with `Orchard` prefix looks consistent.